infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/log/LogWatcher.java

```diff
@@ -0,0 +1,137 @@
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.event.PodEvent;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.event.PodEventHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LogWatcher implements PodEventHandler, Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LogWatcher.class);
+
+  private static final ExecutorService pool = Executors.newFixedThreadPool(10);
+
+  private final String namespace;
+  private final KubernetesClient client;
+  private final PodLogHandler logHandler;
+
+  // set of current watchers. This is used so we're able to cut-off the watchers from outside.
+  private final Set<LogWatch> currentWatchers = new HashSet<>();
+
+  // set of currently observed containers so we're not try to follow same container multiple-times.
+  private final Set<String> watchingContainers = new HashSet<>();
+
+  public LogWatcher(
+      KubernetesClientFactory clientFactory,
+      String workspaceId,
+      String namespace,
+      PodLogHandler handler)
+      throws InfrastructureException {
+    this.logHandler = handler;
+    this.client = clientFactory.create(workspaceId);
+    this.namespace = namespace;
+  }
+
+  @Override
+  public void handle(PodEvent event) {
+    final String podName = event.getPodName();
+
+    if (logHandler.matchPod(podName)) {
+      final String containerName = event.getContainerName();
+      if (containerName != null && !watchingContainers.contains(containerName)) {
+        watchingContainers.add(containerName);
+        LOG.trace(
+            "adding [{}] to watching containers now watching [{}]",
+            containerName,
+            watchingContainers);
+        pool.submit(new ContainerLogWatch(podName, containerName));
+      }
+    }
+  }
+
+  @Override
+  public void close() {
+    new Thread(
+            () -> {
+              try {
+                LOG.debug("Waiting 5s before exit to get all the logs");
+                Thread.sleep(5000);
+              } catch (InterruptedException e) {
+                LOG.error("Interrupted waiting for the logs", e);
+              } finally {
+                currentWatchers.forEach(LogWatch::close);
+                currentWatchers.clear();
+              }
+            })
+        .start();
+  }
+
+  private class ContainerLogWatch implements Runnable {
+
+    private final String podName;
+    private final String containerName;
+
+    private ContainerLogWatch(String podName, String containerName) {
+      this.podName = podName;
+      this.containerName = containerName;
+    }
+
+    /**
+     * Do the best effort to get the logs from the container. The method tries N times to get the
+     * logs from the container. If response on log request from the k8s is 40x, it possibly means
```

Comment:
where does the `40x` come from?

Reply:
response from k8s to watch log request. I've updated the docs. Should I comment it better?